src/base/LruMap.h

```diff
@@ -135,11 +153,29 @@ LruMap<EntryValue, EntryCost>::get(const char *key)
     return NULL;
 }
 
-template <class EntryValue, size_t EntryCost>
+template <class EntryValue, size_t MemoryUsedByEV(const EntryValue *)>
 bool
-LruMap<EntryValue, EntryCost>::add(const char *key, EntryValue *t)
+LruMap<EntryValue, MemoryUsedByEV>::add(const char *key, EntryValue *t)
 {
-    if (ttl == 0)
+    // map the retention concept into a TTL value
+
+    // do not store / retain any entries
+    if (retention == 0)
+        return false;
+
+    // negative retention means store 'forever'
+    if (retention < 0)
+        return add(key, t, 0xFFFFFFFF);
```

Comment:
I think I mentioned it before, but the retention is only known in some cases to the logic instantiating a particular LruMap object. The code using add() is generally unware of it, and in some places cannot even get access to the instantiator logic to find out. Getting the retention info to those latter callers was what abandoned defaultTtl() was all about earlier.